### PR TITLE
Fix factorial2 and pytest fix #308 and #309

### LINKED ISSUE #308: Fix Factorial2
The most recent merged [pull-request](https://github.com/theochem/iodata/pull/301) causes 165 tests to fail in Github [Actions](https://github.com/theochem/iodata/actions/runs/7161627372/job/19497490702#step:6:9878) for various python systems.  The following error is obtained for the vast majority of them:
```python
     def factorial2(n, exact=False):
        """Wrap scipy.special.factorial2 to return 1.0 when the input is -1.
    
        This is a temporary workaround while we wait for Scipy's update.
        To learn more, see https://github.com/scipy/scipy/issues/18409.
    
        Parameters
        ----------
        n : int or np.ndarray
            Values to calculate n!! for. If n={0, -1}, the return value is 1.
            For n < -1, the return value is 0.
        """
        # Scipy  1.11.x returns an integer when n is an integer, but 1.10.x returns an array,
        # so np.array(n) is passed to make sure the output is always an array.
        out = scipy.special.factorial2(np.array(n), exact=exact)
>       out[out <= 0] = 1.0
E       TypeError: 'int' object does not support item assignment
```

Further the following [line](https://github.com/theochem/iodata/blame/6b9c6e0e94a0425593217954f3913f4545a5605d/iodata/overlap.py#L240) does not match [scipy.factorial2](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.factorial2.html) API.  Looking at the earlier release of this package, we want `exact=True` in scipy.factorial2.  When setting `exact=True`, then scipy.factorial2 only accepts integers, and so we will need to run it over each individual element of the array.

The solution would be too in `factorial2` function:
1. Keep `out = scipy.special.factorial2(np.array(n), exact=exact)` when exact=False
2. When exact=True, and 
   i. Add if n is not `int`, then convert it to array `n = np.array([n])`) 
   ii.  compute `scipy.special.factorial2(x, exact)` on each individual element `x` of the array `n`.
   
Thank you to @LudoRNLT and @D-TheProgrammer, if you got the time to look into this.

### PR DESCRIPTION
**Pull request by D-TheProgrammer**
This pull request represents another attempt to address the issue. The previous attempt was unmergeable due to being made from the main branch. The objective of this fix is to address the issues with the `factorial2` function in the `òverlap.py`file  and create a unit test using pytest named `test_factorial2`. Our solution for the function has been validated by passing the linting checks.

### CODE REVIEW

**Comment by D-TheProgrammer** [comment]
@tovrstra Normally it should be good now for the merge let us know 
@ludoRNLT here's the new one i forgot to tag you on this one

**Review by tovrstra — COMMENTED** [review]
@D-TheProgrammer Thanks for the pull request and my apologies for the large number of changes we're going through while you were making pull requests.

There are still a few linters complaining. If you click on "Details" on the "pre-commit.ci - pr" line, and then scroll down, you will see the errors. Normally, when you install pre-commit in the clone of the repository on your computer, it will not allow you to make the commit when there are such issues. If you need help getting that to work, please let me know. (This is the recommended procedure, because fixing issues is always easier right after having written the code.)

If you need help with fixing the linter complaints themselves, just let me know.

File: iodata/overlap.py [hunk 1]
```diff
@@ -25,32 +25,56 @@
...
-    # so np.array(n) is passed to make sure the output is always an array.
-    out = scipy.special.factorial2(np.array(n), exact=exact)
-    out[out <= 0] = 1.0
-    out[out <= -2] = 0.0
-    return out
-
-
+    # Handle integer inputs
+    if isinstance(n, (int, np.integer)):
+        if n in {-1, 0}:
```
Comment (tovrstra) [inline iodata/overlap.py:48]: Creating sets, like `{-1, 0}`, is expensive in Python
```suggestion
        if n in [-1, 0]:
```

File: iodata/overlap.py [hunk 2]
```diff
@@ -25,32 +25,56 @@
...
+            return 1.0
+        if n < -1:
+            return 0.0
+        return scipy.special.factorial2(n, exact=exact)
+
+    # Handle float inputs
+    if isinstance(n, float):
+        if n in {-1.0, 0.0}:
```
Comment (tovrstra) [inline iodata/overlap.py:56]:
```suggestion
        if n in [-1.0, 0.0]:
```

File: iodata/overlap.py [hunk 3]
```diff
@@ -25,32 +25,56 @@
...
+            return 1.0
+        if n < -1.0:
+            return 0.0
+        return scipy.special.factorial2(int(n), exact=exact)
+
+    # Handle array inputs
+    if isinstance(n, np.ndarray):
+        result = np.zeros_like(n, dtype=float)
+        for i, val in np.ndenumerate(n):
+            if val in {-1.0, 0.0}:
```
Comment (tovrstra) [inline iodata/overlap.py:66]:
```suggestion
            if val in [-1.0, 0.0]:
```

**Comment by D-TheProgrammer** [comment]
> @D-TheProgrammer Thanks for the pull request and my apologies for the large number of changes we're going through while you were making pull requests.
> 
> There are still a few linters complaining. If you click on "Details" on the "pre-commit.ci - pr" line, and then scroll down, you will see the errors. Normally, when you install pre-commit in the clone of the repository on your computer, it will not allow you to make the commit when there are such issues. If you need help getting that to work, please let me know. (This is the recommended procedure, because fixing issues is always easier right after having written the code.)
> 
> If you need help with fixing the linter complaints themselves, just let me know.


We tried with our computers and we had 4 or 5 linters that didn't work but they were not tied to our fix it was about the other functions outside of factorial2 but still in overlap.py, this is why we did our PR because we thought that someone else was working on those other issues. 
We

**Comment by tovrstra** [comment]
I have another small question. I may be overlooking something, but I don't believe we need the floating point case in IOData and always need exact=True, which may be used to simplify our version of `factorial2` considerably.

**Comment by tovrstra** [comment]
> @tovrstra We tried with our computers and we had 4 or 5 linters that didn't work but they were not tied to our fix it was about the other functions outside of factorial2 but still in overlap.py, this is why we did our PR because we thought that someone else was working on those other issues. 

Ok, sounds good. No worries. Thanks for the quick reply. I'll take care of it after the weekend

**Comment by tovrstra** [comment]
@D-TheProgrammer I made all improvements I could think of. Can you double-check the final result? Thanks.

**Comment by tovrstra** [comment]
Thanks for checking. I'll merge soon.

**Pull request merged**